### ISSUE TITLE
DNS Lookup Causes Deadlock and Excessive UDP Connections

### ISSUE THREAD

**Issue by javaweb-sec** [dnsjava/dnsjava#326]
## Description
I encountered a deadlock issue while using dnsjava in a multi-threaded environment. During DNS lookup operations, thousands of threads get stuck at the same point for an extended period. This issue also causes a large number of UDP connections to remain unreleased, leading to resource exhaustion.

## Environment
```
Java Version: 21.0.3
Library Version: dnsjava 3.5.3
Operating System: CentOS Linux release 7.6.1810 (Core)
```
## Steps to Reproduce

1. Perform a large number of DNS lookup operations using dnsjava library.
2. Monitor the application threads using jstack.
3. Monitor the UDP connections using netstat and lsof.

## Expected Behavior
. DNS lookup operations should complete successfully, and all UDP connections should be released after the lookup.

## Actual Behavior
Thousands of threads are stuck in the WAITING state, waiting for CompletableFuture.join() to complete.
A large number of UDP connections remain in the ESTABLISHED state and are not released.

## Stack Trace
Example stack trace of stuck threads:
```
"pool-8-thread-1" #470 [1179] prio=5 os_prio=0 cpu=1.69ms elapsed=6013.09s tid=0x00007f60b00032d0 nid=1179 waiting on condition  [0x00007f61586fa000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@21.0.3/Native Method)
	- parking to wait for  <0x00000000ffa4aa48> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(java.base@21.0.3/LockSupport.java:221)
	at java.util.concurrent.CompletableFuture$Signaller.block(java.base@21.0.3/CompletableFuture.java:1864)
	at java.util.concurrent.ForkJoinPool.unmanagedBlock(java.base@21.0.3/ForkJoinPool.java:3780)
	at java.util.concurrent.ForkJoinPool.managedBlock(java.base@21.0.3/ForkJoinPool.java:3725)
	at java.util.concurrent.CompletableFuture.waitingGet(java.base@21.0.3/CompletableFuture.java:1898)
	at java.util.concurrent.CompletableFuture.join(java.base@21.0.3/CompletableFuture.java:2117)
	at org.javaweb.bincat.task.service.DNSLookup.lookupDNS(DNSLookup.java:83)
	...
```

## UDP Connections
Example of excessive UDP connections:

```
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
udp        0      0 172.31.101.190:41893    1.1.1.1:53              ESTABLISHED
udp        0      0 172.31.101.190:48037    1.1.1.1:53              ESTABLISHED
...
```

## lsof -p xxx
```
java    1142 root *993u     IPv4          154730308       0t0       UDP iZj6ca8k1p60k72c9acavxZ:43463->one.one.one.one:domain 
java    1142 root *994u     IPv4          154730309       0t0       UDP iZj6ca8k1p60k72c9acavxZ:40574->one.one.one.one:domain 
java    1142 root *995u     IPv4          154730310       0t0       UDP iZj6ca8k1p60k72c9acavxZ:43481->one.one.one.one:domain 
java    1142 root *996u     IPv4          154733336       0t0       UDP iZj6ca8k1p60k72c9acavxZ:51107->one.one.one.one:domain 
java    1142 root *997u     IPv4          154733337       0t0       UDP iZj6ca8k1p60k72c9acavxZ:48628->one.one.one.one:domain 
java    1142 root *998u     IPv4          154733338       0t0       UDP iZj6ca8k1p60k72c9acavxZ:51666->one.one.one.one:domain
...
```

## Code

```java
@Bean
public Resolver createResolver() {
// Create an array of Resolver objects from the configured DNS servers
Resolver[] resolvers = Arrays.stream(dnsServers)
		.map(server -> {
			try {
				// Create a SimpleResolver for each DNS server
				SimpleResolver resolver = new SimpleResolver(server);

				// Set the timeout for each resolver
				resolver.setTimeout(Duration.ofMillis(dnsTimeout));

				return resolver;
			} catch (UnknownHostException e) {
				// Log any errors encountered while creating resolvers
				logger.info("Error creating resolver for {}: {}", server, e.getMessage());

				return null;
			}
		})
		// Remove any null resolvers (those that failed to create)
		.filter(Objects::nonNull)
		.toArray(Resolver[]::new);

// If no valid resolvers were created, use the default resolver
if (resolvers.length == 0) {
	logger.info("No valid resolvers created. Using default resolver.");

	try {
		return new SimpleResolver();
	} catch (UnknownHostException e) {
		throw new RuntimeException("Failed to create default resolver", e);
	}
}

return new ExtendedResolver(resolvers);
}

@Bean
public LookupSession createLookupSession(Resolver resolver) {
return LookupSession.builder()
		.resolver(resolver)
		.build();
}
```
**DNSLookup.java:**

```java

public class DNSLookup {

	private LookupSession session;

	@Autowired
	public void setSession(LookupSession session) {
		this.session = session;
	}

	/**
	 * Asynchronously lookup DNS records for a specified domain name.
	 *
	 * @param domain The domain name to lookup
	 * @param types  The DNS record types to query (e.g., A, AAAA, MX)
	 * @return A CompletableFuture containing the lookup results
	 */
	public Set<Map<String, Object>> lookupDNS(String domain, int[] types) {
		if (types.length == 0) {
			return Set.of();
		}

		Map<String, CompletableFuture<LookupResult>> lookups = new LinkedHashMap<>();

		boolean isValidDomain;
		boolean isTopDomain = false;
		boolean isValidIP   = NetworkUtils.isValidIP(domain);

		if (!isValidIP) {
			isValidDomain = isValidDomain(domain);

			if (isValidDomain) {
				isTopDomain = isTopDomain(domain);
			}
		}

		// Perform async DNS lookup for each type
		for (int type : types) {
			// IP only allows ptr query
			if (isValidIP && type != Type.PTR) {
				continue;
			}

			// The root domain does not allow cname queries
			if (type == Type.CNAME && isTopDomain) {
				continue;
			}

			// Subdomains do not allow cname queries
			if (type == Type.NS && !isTopDomain) {
				continue;
			}

			lookups.put(Type.string(type), lookupAsync(domain, type));
		}

		// Wait for all async lookups to complete, then process the results
		return parseRecords(domain, CompletableFuture.allOf(lookups.values().toArray(new CompletableFuture[0]))
				.thenApply(v -> {
					Map<String, List<Record>> result = new HashMap<>();

					lookups.forEach((key, future) -> result.put(key, future.join().getRecords()));

					return result;
				})
				.exceptionally(ex -> Map.of()).join());
	}

	/**
	 * Asynchronously look up DNS records for a given domain name.
	 *
	 * @param domain The domain name to look up.
	 * @param type   The type of DNS record to look up.
	 * @return A CompletableFuture containing the lookup result.
	 */
	public CompletableFuture<LookupResult> lookupAsync(String domain, int type) {
		Name name = Name.fromConstantString(domain);

		// If the type is PTR, convert the domain to a reverse lookup address.
		if (type == Type.PTR) {
			try {
				name = ReverseMap.fromAddress(domain);
			} catch (UnknownHostException ignored) {
				// Ignore exception
			}
		}

		// Perform the asynchronous lookup and return a CompletableFuture.
		return session.lookupAsync(name, type).toCompletableFuture();
	}

	/**
	 * Parse DNS records and convert them into a set of maps containing key-value pairs.
	 *
	 * @param domain  The domain name associated with the records.
	 * @param records A map containing DNS records.
	 * @return A set of maps representing the parsed DNS records.
	 */
	public Set<Map<String, Object>> parseRecords(String domain, Map<String, List<Record>> records) {
		Set<Map<String, Object>> list = new HashSet<>();

		// Iterate through each record type and its corresponding list of records.
		records.forEach((k, v) -> v.forEach(record -> {
			Map<String, Object> recordMap = new LinkedHashMap<>();

			if (record instanceof CNAMERecord cnameRecord) {
				recordMap.put("key", cnameRecord.getName().toString(true));
				recordMap.put("value", cnameRecord.getTarget().toString(true));
			} else if (record instanceof ARecord aRecord) {
				String name  = aRecord.getName().toString(true);
				String value = aRecord.getAddress().getHostAddress();
				recordMap.put("value", value);

				// CNAME
				if (!domain.equals(name)) {
					list.add(Map.of("key", "CNAME", "value", name));
				}
			} else if (record instanceof AAAARecord aaaaRecord) {
				recordMap.put("value", aaaaRecord.getAddress().getHostAddress());
			} else if (record instanceof MXRecord mxRecord) {
				recordMap.put("value", mxRecord.getTarget().toString(true));
			} else if (record instanceof NSRecord nsRecord) {
				recordMap.put("value", nsRecord.getTarget().toString(true));
			} else if (record instanceof PTRRecord ptrRecord) {
				String ptr = ptrRecord.getTarget().toString(true);
				recordMap.put("value", ptr != null ? ptr : domain);
			}

			if (!recordMap.isEmpty()) {
				recordMap.put("type", Type.string(record.getType()));
				list.add(recordMap);
			}
		}));

		return list;
	}

}
```

**Worker Thread**
```java
try (ExecutorService executor = Executors.newVirtualThreadPerTaskExecutor()) {
for (int i = 0; i < poolSize; i++) {
	executor.execute(() -> {
		Object obj;

		while ((obj = queue.poll()) != null) {
			if (obj instanceof String line) {
				long index = COUNTER.incrementAndGet();

				if (index % 500 == 0) {
					logger.info("current index:" + index);
				}

				task.execute(line);
			}
		}
	});
}

executor.shutdown();

while (!executor.awaitTermination(1, TimeUnit.SECONDS)) {
	logger.debug("Waiting for tasks in the thread pool to complete...");
}
```

**DNS query Thread**
```java
public void execute(String domain) {
		try (ExecutorService executor = Executors.newSingleThreadExecutor()) {
			Future<?> future = executor.submit(() -> {
				Set<Map<String, Object>> records = dnsLookup.lookupDNS(domain, new int[]{A, AAAA, NS, MX});

				if (!records.isEmpty()) {
					binCatTaskManager.writeData(JSON.toJSONString(Map.of("domain", domain, "records", records)));
				}
			});

			try {
				// dnsTimeout + 1000
				future.get(dnsTimeout + TimeUnit.SECONDS.toMillis(1), TimeUnit.MILLISECONDS);
			} catch (Exception e) {
				logger.debug("The domain {} DNS parse timed out and has been forcibly closed.", domain);
				future.cancel(true);
			}
		}
	}
```

**Comment by ibauersachs** [comment]
Thanks for the detailed report. Unfortunately the stacktrace doesn't show the frames from dnsjava. Can you please check if this isn't a duplicate of #315 with a build from the current master branch?

**Comment by javaweb-sec** [comment]
I saw the same blocking chain in issue #315 . It seems to be the same problem, but further confirmation is needed.

Below is a partial call chain from my jstack:
```java
"ForkJoinPool.commonPool-worker-1" #3105 [1275] daemon prio=5 os_prio=0 cpu=92.63ms elapsed=6012.84s tid=0x00007f615c004170 nid=1275 waiting for monitor entry  [0x00007f609ad43000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.xbill.DNS.NioClient.setCloseTask(NioClient.java:149)
	- waiting to lock <0x00000000ffa9dd70> (a java.lang.Class for org.xbill.DNS.NioClient)
	at org.xbill.DNS.NioTcpClient.<clinit>(NioTcpClient.java:33)
	at org.xbill.DNS.SimpleResolver.sendAsync(SimpleResolver.java:371)
	at org.xbill.DNS.SimpleResolver.lambda$sendAsync$1(SimpleResolver.java:446)
	at org.xbill.DNS.SimpleResolver$$Lambda/0x00007f610c326230.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(java.base@21.0.3/CompletableFuture.java:1150)
	at java.util.concurrent.CompletableFuture$Completion.exec(java.base@21.0.3/CompletableFuture.java:483)
	at java.util.concurrent.ForkJoinTask.doExec(java.base@21.0.3/ForkJoinTask.java:387)
	at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(java.base@21.0.3/ForkJoinPool.java:1312)
	at java.util.concurrent.ForkJoinPool.scan(java.base@21.0.3/ForkJoinPool.java:1843)
	at java.util.concurrent.ForkJoinPool.runWorker(java.base@21.0.3/ForkJoinPool.java:1808)
	at java.util.concurrent.ForkJoinWorkerThread.run(java.base@21.0.3/ForkJoinWorkerThread.java:188)

"pool-99-thread-1" #3134 [1276] prio=5 os_prio=0 cpu=1.11ms elapsed=6012.83s tid=0x00007f60ac029130 nid=1276 waiting on condition  [0x00007f609ac43000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@21.0.3/Native Method)
	- parking to wait for  <0x00000000ffabe150> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(java.base@21.0.3/LockSupport.java:221)
	at java.util.concurrent.CompletableFuture$Signaller.block(java.base@21.0.3/CompletableFuture.java:1864)
	at java.util.concurrent.ForkJoinPool.unmanagedBlock(java.base@21.0.3/ForkJoinPool.java:3780)
	at java.util.concurrent.ForkJoinPool.managedBlock(java.base@21.0.3/ForkJoinPool.java:3725)
	at java.util.concurrent.CompletableFuture.waitingGet(java.base@21.0.3/CompletableFuture.java:1898)
	at java.util.concurrent.CompletableFuture.join(java.base@21.0.3/CompletableFuture.java:2117)
	at org.javaweb.bincat.task.service.DNSLookup.lookupDNS(DNSLookup.java:83)
	at org.javaweb.bincat.task.executor.DNSTaskExecutor.lambda$execute$0(DNSTaskExecutor.java:36)
	at org.javaweb.bincat.task.executor.DNSTaskExecutor$$Lambda/0x00007f610c30e220.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(java.base@21.0.3/Executors.java:572)
	at java.util.concurrent.FutureTask.run(java.base@21.0.3/FutureTask.java:317)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@21.0.3/ThreadPoolExecutor.java:1144)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@21.0.3/ThreadPoolExecutor.java:642)
	at java.lang.Thread.runWith(java.base@21.0.3/Thread.java:1596)
	at java.lang.Thread.run(java.base@21.0.3/Thread.java:1583)
	```

**Comment by javaweb-sec** [comment]
I saw the same blocking chain in issue #315 . It seems to be the same problem, but further confirmation is needed.

Below is a partial call chain from my jstack:
```java
"ForkJoinPool.commonPool-worker-1" #3105 [1275] daemon prio=5 os_prio=0 cpu=92.63ms elapsed=6012.84s tid=0x00007f615c004170 nid=1275 waiting for monitor entry  [0x00007f609ad43000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.xbill.DNS.NioClient.setCloseTask(NioClient.java:149)
	- waiting to lock <0x00000000ffa9dd70> (a java.lang.Class for org.xbill.DNS.NioClient)
	at org.xbill.DNS.NioTcpClient.<clinit>(NioTcpClient.java:33)
	at org.xbill.DNS.SimpleResolver.sendAsync(SimpleResolver.java:371)
	at org.xbill.DNS.SimpleResolver.lambda$sendAsync$1(SimpleResolver.java:446)
	at org.xbill.DNS.SimpleResolver$$Lambda/0x00007f610c326230.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(java.base@21.0.3/CompletableFuture.java:1150)
	at java.util.concurrent.CompletableFuture$Completion.exec(java.base@21.0.3/CompletableFuture.java:483)
	at java.util.concurrent.ForkJoinTask.doExec(java.base@21.0.3/ForkJoinTask.java:387)
	at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(java.base@21.0.3/ForkJoinPool.java:1312)
	at java.util.concurrent.ForkJoinPool.scan(java.base@21.0.3/ForkJoinPool.java:1843)
	at java.util.concurrent.ForkJoinPool.runWorker(java.base@21.0.3/ForkJoinPool.java:1808)
	at java.util.concurrent.ForkJoinWorkerThread.run(java.base@21.0.3/ForkJoinWorkerThread.java:188)

"pool-99-thread-1" #3134 [1276] prio=5 os_prio=0 cpu=1.11ms elapsed=6012.83s tid=0x00007f60ac029130 nid=1276 waiting on condition  [0x00007f609ac43000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@21.0.3/Native Method)
	- parking to wait for  <0x00000000ffabe150> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(java.base@21.0.3/LockSupport.java:221)
	at java.util.concurrent.CompletableFuture$Signaller.block(java.base@21.0.3/CompletableFuture.java:1864)
	at java.util.concurrent.ForkJoinPool.unmanagedBlock(java.base@21.0.3/ForkJoinPool.java:3780)
	at java.util.concurrent.ForkJoinPool.managedBlock(java.base@21.0.3/ForkJoinPool.java:3725)
	at java.util.concurrent.CompletableFuture.waitingGet(java.base@21.0.3/CompletableFuture.java:1898)
	at java.util.concurrent.CompletableFuture.join(java.base@21.0.3/CompletableFuture.java:2117)
	at org.javaweb.task.service.DNSLookup.lookupDNS(DNSLookup.java:83)
	at org.javaweb.task.executor.DNSTaskExecutor.lambda$execute$0(DNSTaskExecutor.java:36)
	at org.javaweb.task.executor.DNSTaskExecutor$$Lambda/0x00007f610c30e220.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(java.base@21.0.3/Executors.java:572)
	at java.util.concurrent.FutureTask.run(java.base@21.0.3/FutureTask.java:317)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@21.0.3/ThreadPoolExecutor.java:1144)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@21.0.3/ThreadPoolExecutor.java:642)
	at java.lang.Thread.runWith(java.base@21.0.3/Thread.java:1596)
	at java.lang.Thread.run(java.base@21.0.3/Thread.java:1583)
	```

**Comment by ibauersachs** [comment]
v3.6.0 is released now and I assume the issue is solved there. We can always reopen this issue if it persist in 3.6.0.
